Test StringHierarchyRegion

cld99 opened the pull request.

cld99 committed:
Create testing class for StringHierarchyRegion

cld99 committed:
Add tests for constructor null parameters

lhmcgann commented:
Just edit this PR to merge back onto release-v2.0 rather than main branch

cld99 committed:
Add remaining test cases

cld99 committed:
Finalize test cases and add null argument checking to AddSubregion

lhmcgann commented on EStimLibrary.sln:
```diff
@@ -5,23 +5,26 @@ VisualStudioVersion = 17.0.31903.59
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "src", "src", "{7E35BD38-1080-4A2E-B8B4-7CB0C5C08607}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "EStimLibrary", "src\EStimLibrary\EStimLibrary.csproj", "{23710686-EDC8-41FC-AAC8-52B002F488D6}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "EStimLibrary", "src\EStimLibrary\EStimLibrary.csproj", "{23710686-EDC8-41FC-AAC8-52B002F488D6}"
```
Revert changes to the EStimLibrary.sln file. You should be able to use the tests/EStimLibrary.UnitTests/EStimLibrary.UnitTests.sln to open both the library and unit test projects at once. EStimLibrary.sln is just for the library

cld99 replied:
When I was initially trying to open the project in Visual Studio, it automatically opened the EStimLibrary.sln file, and I was initially trying to make changes from there. I later realized that it had opened the wrong solution and opened it correctly, but it looks like the changes were committed. I will revert them.

lhmcgann commented on src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyRegion.cs:
```diff
@@ -1,4 +1,6 @@
-﻿namespace EStimLibrary.Extensions.SpatialModel.StringHierarchy;
+﻿using System;
```
were you getting an error for not having this? It should be found via the global Usings.cs file already

cld99 replied:
No, I was not getting an error. I don't remember adding this, so I will just remove it.

lhmcgann commented on src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyRegion.cs:
```diff
@@ -120,6 +122,10 @@ public StringHierarchyRegion(string baseName, StringHierarchyRegion parent,
     public void AddSubregion(StringHierarchyRegion subregion,
         out StringHierarchyRegion? existingSubregion)
     {
+        /// <exception cref = "ArgumentNullException"> The provided 
```
Just curious: when did you run into this case?

cld99 replied:
In theory, you should never run into this case if the method is used properly, but I wanted to have explicit error handling since the method attempts to access fields of the subregion.

lhmcgann committed:
Comments and formatting

lhmcgann commented on src/EStimLibrary/Extensions/SpatialModel/StringHierarchy/StringHierarchyRegion.cs:
```diff
@@ -117,9 +118,16 @@ public StringHierarchyRegion(string baseName, StringHierarchyRegion parent,
     public void AddSubregion(StringHierarchyRegion subregion,
         out StringHierarchyRegion? existingSubregion)
     {
+        if (subregion == null)
```
Just curious: when did you run into this case?

lhmcgann replied:
??

cld99 replied:
As per my response to your Nov. 18th comment, "In theory, you should never run into this case if the method is used properly, but I wanted to have explicit error handling since the method attempts to access fields of the subregion".

lhmcgann commented on tests/EStimLibrary.UnitTests/Extensions/SpatialModel/StringHierarchy/StringHierarchyRegionTests.cs:
```diff
@@ -0,0 +1,547 @@
+    [Fact]
+    public void Constructor_ShouldAcceptNull()
+    {
+        var stringHierarchyRegion = new StringHierarchyRegion("base", null,
```
Nice test cases from a first quick look! Would you mind just formatting them - i.e., all lines <=80 characters, kinda like I've done in the first few sections here - and adding some in-line comments throughout your method bodies to help me quickly get a grasp of what you're testing in each? Both what structures you're making and then what you're asserting?

johnshi314 reviewed (dismissed):
I don't see any issues here, so it should be good to go.

cld99 committed:
Revert changes to EStimLibrary.sln

cld99 committed:
Format testing class and add comments

cld99 committed:
Resolve merge conflict

lhmcgann commented on docs/images/association_arrows-techvidvan.png:
Are you by chance able to undo these changes to the image files? They have already been pushed in release which I know you aren't merging into yet, so if not possible it's alright we'll deal with it later, but if you can it will help prevent merge conflicts later.

lhmcgann reviewed (changes requested):
Looking good! So almost there! Just a couple of in-line review comments and one overall one: make sure to check the test coverage report of the StringHierarchyRegion class and add test cases to cover the lines it highlights as not being covered yet. you can find the coverage report on the latest pushed commit that changed source code or build files. Under the commit you should see a little green checkmark and a 3/3. Click on that and then go to the Details of any of the 3 OS builds. Expand the output for the step title "Upload coverage report" or similar and click the download link for the report. Then in the downloaded folder, open the .HTML file of the StringHierarchyRegion class

cld99 committed:
Fix code coverage for StringHierarchyRegionTests

cld99 committed:
Merge branch 'release-v2.0' into test/string-hierarchy-region

cld99 committed:
Remove extraneous error handling

lhmcgann committed:
Two bugs found

1. IsValidModifierSpec: shared modifiers don't work. See TODO comments.
2. DeepCopy: parent options weren't copied (easy fix). New test method scheme.

Comments in source class.

lhmcgann committed:
Added comments

lhmcgann committed:
Added tests; commenting

lhmcgann reviewed (approved):
@cld99 nice job! Just to get this one done, I added doc comments in the source and test class, and a few more test cases. Some of the new edge cases brought up a couple bugs: one was an easy fix, one is more challenging and I need to think on a good solution approach. For the test cases I added, I mostly used the [MemberData()] feature to make it really easy to run the same Assert checks on different inputs. I also made some static Construct<TestRootRegion> functions to help with not needing to re-init new test data structs inside each test method. Check out the changes as a reference for next time! Merging this one in now though

lhmcgann committed:
Merge branch 'release-v2.0' into test/string-hierarchy-region
